File: www/chromiumcu/chromiumcu.js
/*
 * Copyright 2017 Kopano and its licensors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

'use strict';

import { JanusVideoCall } from './plugins/janus.js';

export class ChromiuMCU {
	constructor(uri='ws://localhost:8778/api/v1/mcu/websocket', handler=null, options=null) {
		console.debug('new chromiumcu', this);

		this.uri = uri;
		this.handler = handler;

		this.ws = null;
		this.reconnectEnabled = handler === null ? true : false;
		this.reconnectInterval = 1000;
		this.maxReconnectInterval = 30000;
		this.reconnectFactor = 1.5;
		this.connectTimeout = 5000;
		this.reconnectAttempts = 0;

		this.plugins = {};

		if (!options) {
			options = {};
		}

		[JanusVideoCall].forEach(plugin => {
			this.plugins[plugin.plugin] = plugin;
			console.info('registered plugin', plugin.plugin);
		});

		Object.entries(options).forEach(([key, value]) => {
			this[key] = value;
		});
	}

	autostart(query='', hash='', reconnect=true) {
		return new Promise((resolve, reject) => {
			const queryParams = decodeParams(query);
			const hashParams = decodeParams(hash);
			console.info('chromiumcu autostart', queryParams, hashParams);

			if (queryParams.uri) {
				this.uri = queryParams.uri;
			}
			if (hashParams.plugin) {
				console.log('chromiumcu launching plugin', hashParams.plugin);
				let plugin = this.plugins[hashParams.plugin];
				if (!plugin) {
					reject(new Error('mcu unknown plugin: ' + hashParams.plugin));
					return;
				}
				this.handler = new plugin();
				if (hashParams.id) {
					this.uri = this.uri + '/' + hashParams.id;
				}
			}

			this.connect(reconnect).then(socket => {
				resolve(socket);
			}).catch(err => {
				reject(err);
			});
		});
	}

	connect(reconnect=false) {
		return new Promise((resolve, reject) => {
			console.info('connecting to', this.uri, reconnect);
			let connected = false;
			let timedout = false;
			if (reconnect && !this.reconnectEnabled) {
				reconnect = false;
			}

			let ws = new WebSocket(this.uri, 'kwmmcu-protocol');

			let reconnector = () => {
				if (!connected && !reconnect) {
					return;
				}
				if (this.handler !== null) {
					// For now never reconnect handler connections.
					return;
				}
				let reconnectDelay = this.reconnectInterval * Math.pow(this.reconnectFactor, this.reconnectAttempts);
				setTimeout(() => {
					this.reconnectAttempts++;
					this.connect(true);
				}, reconnectDelay > this.maxReconnectInterval ? this.maxReconnectInterval : reconnectDelay);
				console.info('websocket reconnecting in', reconnectDelay);
			};

			let timeout = setTimeout(() => {
				ws.close();
				if (reconnect) {
					reconnector();
					return;
				}

				reject(new Error('connection timeout'));
				if (this.ws !== ws) {
					return;
				}
				this.ws = null;
			}, this.connectTimeout);

			ws.onclose = event => {
				clearTimeout(timeout);
				if (this.ws !== ws) {
					return;
				}
				console.warn('websocket close', event);
				if (this.handler !== null && this.handler.onclose) {
					this.handler.onclose(event);
				}
				this.ws = null;
				reconnector();
			};
			ws.onerror = err => {
				if (timedout) {
					return;
				}
				console.error('websocket error', err);
				if (this.handler !== null && this.handler.onerror) {
					this.handler.onerror(err);
				}
				if (reconnect && !connected) {
					return;
				}
				reject(err);
			};
			ws.onmessage = event => {
				if (this.handler !== null) {
					this.handler.process(event.target, JSON.parse(event.data));
				} else {
					this.process(JSON.parse(event.data));
				}
			};
			ws.onopen = event => {
				clearTimeout(timeout);
				connected = true;
				if (timedout) {
					return;
				}
				console.info('websocket open', event);
				if (this.handler !== null && this.handler.onopen) {
					this.handler.onopen(event);
				}
				resolve(ws);
				this.reconnectAttempts = 0;
			};

			this.ws = ws;
		});
	}

	process(msg) {
		console.debug('processing mcu msg', msg);
		switch (msg.type) {
			case 'attach':
				let plugin = this.plugins[msg.plugin];
				if (!plugin) {
					console.warn('mcu unknown plugin', msg.plugin);
					return;
				}

				// Create iframe and load handler in there.
				let iframe = document.createElement('iframe');
				iframe.onload = event => {
					console.debug('iframe loaded', event);
				};
				iframe.src = location.href + '#plugin=' + msg.plugin + '&id=' + msg.transaction;
				document.body.appendChild(iframe);
				break;

			default:
				console.warn('mcu unknown type', msg.type);
				break;
		}
	}
}


function decodeParams(s) {
	var regex = /([^&=]+)=([^&]*)/g;
	var result = {};
	var m;
	while (m = regex.exec(s)) {
		result[decodeURIComponent(m[1])] = decodeURIComponent(m[2]);
	}

	return result;
}

File: signaling/janus/plugins/chromiumcu.go
/*
 * Copyright 2017 Kopano and its licensors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package plugins

import (
	"encoding/json"
	"fmt"
	"sync"

	"github.com/sirupsen/logrus"

	"stash.kopano.io/kwm/kwmserver/signaling/api-v1/mcu"
	"stash.kopano.io/kwm/kwmserver/signaling/connection"
	"stash.kopano.io/kwm/kwmserver/signaling/janus"
)

const (
	pluginNameChromiuMCU = "janus.plugin.kopano-chromiumcu"
)

type pluginChromiuMCU struct {
	sync.RWMutex

	name     string
	handleID int64

	logger logrus.FieldLogger

	mcum  *mcu.Manager
	janus *janus.Manager

	connection          *connection.Connection
	connecting          bool
	onAttachedCallbacks []func(janus.Plugin)
}

func newPluginChromiuMCU(name string, handleID int64, mcum *mcu.Manager, janusManager *janus.Manager) janus.Plugin {
	return &pluginChromiuMCU{
		name:     name,
		handleID: handleID,

		logger: janusManager.Logger().WithField("handle_id", handleID),

		mcum:  mcum,
		janus: janusManager,
	}
}

// ChromiuMCUFactory returns the factory function to create new chromiuMCU plugins.
func ChromiuMCUFactory(mcum *mcu.Manager) (string, func(string, *janus.Manager) (janus.Plugin, error)) {
	return pluginNameChromiuMCU, func(name string, m *janus.Manager) (janus.Plugin, error) {
		return newPluginChromiuMCU(name, m.NewHandle(), mcum, m), nil
	}
}

func (p *pluginChromiuMCU) Name() string {
	return p.name
}

func (p *pluginChromiuMCU) HandleID() int64 {
	return p.handleID
}

func (p *pluginChromiuMCU) Logger() logrus.FieldLogger {
	return p.logger
}

func (p *pluginChromiuMCU) OnMessage(m *janus.Manager, c *connection.Connection, msg *janus.MessageMessageData) error {
	p.Lock()
	mcu := p.connection
	p.Unlock()
	if mcu == nil {
		p.Logger().Errorf("no mcu connection available: %s", []byte(*msg.Body))
		return fmt.Errorf("no mcu connection")
	}

	return mcu.SendTransaction(msg, func(reply []byte) error {
		//p.Logger().Debugf("received transaction reply %s, %v", reply, c.ID())
		return p.OnReply(c, reply)
	})
}

func (p *pluginChromiuMCU) OnReply(c *connection.Connection, msg []byte) error {
	// TODO(longsleep): Reuse msg []byte slices / put into pool.
	var envelope janus.EnvelopeData
	err := json.Unmarshal(msg, &envelope)
	if err != nil {
		return err
	}

	if envelope.TargetSession == 0 {
		if envelope.TransactionID() == "" {
			return fmt.Errorf("no target session in reply without transaction")
		}
		return c.RawSend(msg)
	}

	connection := p.janus.GetConnectionBySessionID(envelope.TargetSession)
	if connection == nil {
		return fmt.Errorf("no connection for target session in reply without transaction")
	}

	return connection.RawSend(msg)
}

func (p *pluginChromiuMCU) OnText(c *connection.Connection, msg []byte) error {
	return p.OnReply(c, msg)
}

func (p *pluginChromiuMCU) OnDetach(m *janus.Manager, c *connection.Connection, msg *janus.EnvelopeData) error {
	return nil
}

func (p *pluginChromiuMCU) Attach(m *janus.Manager, c *connection.Connection, msg *janus.AttachMessageData, cb func(janus.Plugin), cleanup func(janus.Plugin)) error {
	p.Lock()
	if p.connection != nil || p.connecting {
		p.Unlock()
		return p.onAttached(m, c, msg, cb, cleanup)
	}
	p.connecting = true
	p.Unlock()

	p.Logger().Debugln("attaching with mcu")
	mcu, err := p.mcum.Attach(p.Name(), p.HandleID(), func(conn *connection.Connection) error {
		// connect callback.
		p.Lock()
		if !p.connecting {
			p.Unlock()
			conn.Close()
			errAttaching := fmt.Errorf("lost mcu connection while attaching")
			p.Logger().Debugln(errAttaching)
			return errAttaching
		}

		p.Logger().WithField("mcu_connection", conn.ID()).Debugln("attached with mcu")
		p.connection = conn
		p.connecting = false
		onAttachedCallbacks := p.onAttachedCallbacks
		p.onAttachedCallbacks = nil
		p.Unlock()
		conn.OnClosed(func(closedConnection *connection.Connection) {
			// closed callback
			p.Logger().WithField("mcu_connection", closedConnection.ID()).Debugln("attached mcu connection(a) has closed")
			p.onClosed()
			if cleanup != nil {
				cleanup(p)
			}
		})

		go func() {
			if cb != nil {
				cb(p)
			}
			for _, attachedCb := range onAttachedCallbacks {
				attachedCb(p)
			}
		}()
		return nil
	}, p.OnText)
	if err != nil {
		p.Logger().Errorf("no mcu connection available for attaching")
		return err
	}

	mcu.OnClosed(func(conn *connection.Connection) {
		// closed callback.
		p.Logger().WithField("mcu_connection", conn.ID()).Warnln("mcu connection has closed")
		p.Lock()
		if p.connection != nil {
			p.connection.RawSend(nil) // This closes.
		} else {
			p.connecting = false
		}
		p.Unlock()
	})

	return nil
}

func (p *pluginChromiuMCU) onAttached(m *janus.Manager, c *connection.Connection, msg *janus.AttachMessageData, cb func(janus.Plugin), cleanup func(janus.Plugin)) error {
	if cb == nil && cleanup == nil {
		return nil
	}

	p.Lock()
	attachedConnection := p.connection
	if attachedConnection == nil && !p.connecting {
		p.Unlock()
		return p.Attach(m, c, msg, cb, cleanup)
	}

	if cleanup != nil {
		if attachedConnection != nil {
			attachedConnection.OnClosed(func(closedConnection *connection.Connection) {
				// closed callback.
				p.Logger().WithField("mcu_connection", closedConnection.ID()).Debugln("attached mcu  connection(b) has closed ")
				p.onClosed()
				if cleanup != nil {
					cleanup(p)
				}
			})
		} else {
			p.onAttachedCallbacks = append(p.onAttachedCallbacks, func(_ janus.Plugin) {
				// attached callback.
				p.RLock()
				attachedConnection2 := p.connection
				p.RUnlock()
				attachedConnection2.OnClosed(func(closedConnection *connection.Connection) {
					// closed callback.
					p.Logger().WithField("mcu_connection", closedConnection.ID()).Debugln("attached mcu  connection(c) has closed")
					p.onClosed()
					if cleanup != nil {
						cleanup(p)
					}
				})
			})
		}
	}

	if cb != nil {
		if attachedConnection != nil {
			p.Unlock()
			go cb(p)
		} else {
			p.onAttachedCallbacks = append(p.onAttachedCallbacks, cb)
			p.Unlock()
		}
	} else {
		p.Unlock()
	}

	return nil
}

func (p *pluginChromiuMCU) onClosed() {
	p.Lock()
	p.connection = nil
	p.Unlock()
}

func (p *pluginChromiuMCU) Enabled() bool {
	p.RLock()
	defer p.RUnlock()
	return p.connection != nil || p.connecting
}
